PARTIAL FIX: Improve leading zeros preservation with dtype=str for dict-based dtypes

dxdc opened the pull request:
## Summary
This PR partially addresses issue #57666 by improving leading zeros preservation when `dtype=str` is used with dictionary-based dtype specifications. While the global `dtype=str` issue with pyarrow engine remains unfixed, this PR resolves the problem for more targeted dtype specifications.

## Problem
Issue #57666 reports that the pyarrow engine does not preserve leading zeros in numeric-looking strings when `dtype=str` is specified, while other engines correctly preserve them.

## Solution
- **Fixed**: Dictionary-based dtype specifications (`dtype={'col': str}`) now properly preserve leading zeros across all engines
- **Partial**: Global `dtype=str` still fails with pyarrow engine (marked with xfail for now)
- **Added**: Test coverage for dtype specification patterns

## What's Fixed vs Still Broken

### ✅ Now Working:
```python
# This now preserve leading zeros correctly across all engines:
pd.read_csv(data, dtype={'col2': str, 'col3': int, 'col4': str})
```

### ⚠️ Still Broken (pyarrow only):
```python
# This still strips leading zeros with pyarrow engine:
pd.read_csv(data, dtype=str)  # global string dtype
```

## Next Steps
This PR provides a foundation for the complete fix. The remaining work involves:
1. Fully resolving the pyarrow engine's global `dtype` handling 
2. Removing the xfail marker once completely resolved
3. Improving the pyarrow engine's dtype enforcement during parsing rather than post-processing conversion

---

## Checklist

- [x] Tests added and passed
- [x] All code checks passed  
- [x] closes #57666
- [x] Added entry in `doc/source/whatsnew/v3.0.0.rst`

## Files Changed
- `pandas/io/parsers/arrow_parser_wrapper.py` - Fix for dict-based dtypes
- `pandas/tests/io/parser/test_preserve_leading_zeros.py` - Comprehensive test suite

## Test Output
- **C engine**: ✅ All tests pass  
- **Python engine**: ✅ All tests pass
- **PyArrow engine**: 
  - Dict-based dtypes now pass (with strings)
  - ⚠️ Global `dtype=str` marked as xfail (temporary)

dxdc committed:
BUG: Preserve leading zeros with dtype=str in pyarrow engine (#57666)

dxdc commented:
@jbrockmendel test is passing now. I have it marked as `xfail` for pyarrow only, but you can clearly see the issue. Once the issue is remedied we can remove the try/except block.

jbrockmendel commented on pandas/tests/io/parser/test_preserve_leading_zeros.py:
```diff
@@ -0,0 +1,43 @@
+        assert result.loc[2, "col4"] == "0205", "lost zeros in col4 row 2"
+
+    except AssertionError as exc:
+        if engine_name == "pyarrow":
```
take a look at how we handle xfails elsewhere.  we check and add the marker before the meat of the test

jbrockmendel replied:
> Is there another preferred way to handle this?

No, just the one I already told you to use.

dxdc commented on pandas/tests/io/parser/test_preserve_leading_zeros.py:
```diff
@@ -0,0 +1,43 @@
+import pytest
+
+
+def test_leading_zeros_preserved_with_dtype_str(all_parsers, request):
```
It seemsl ike the `tests/io/parser` is the right place, but I don't see any other files there that seem appropriate. could you suggest another place? happy to move it.

dxdc commented:
> big picture adding a test isn't wrong, but not a good use of time. if you'd like to actually fix the issue, i think there are some good comments in the original thread

I agree, but the fix doesn't appear to be very straightforward. Happy to work on it if there is some guidance on where to find the relevant pieces. It requires proper mapping of pandas dtypes to pyarrow types, and also handling other logic that pandas supports but pyarrow doesn't (e.g., col index-based dtypes, global dtypes, etc.).

dxdc commented:
@jbrockmendel I dug into the pandas and PyArrow APIs and landed on a more general fix for the issue. I wasn't sure how to run `pytest` locally against the dev branch - my attempts with the Docker container didn't work - so I relied on pandas' test suite.

This patch improves dtype handling during the pyarrow write path by converting supported column-specific dtypes into pyarrow types and passing them via convert_options["column_types"]. However, a few limitations remain:

### Known Issues / Remaining Work

1. Global dtype support is not implemented
The current logic handles only column-specific dtype dictionaries. Global dtypes (e.g., dtype=str) are ignored, which could lead to inconsistent behavior across engines, especially with things like leading-zero string preservation. Supporting this would require column name/index context, which doesn't seem to be readily available here. I couldn't find a safe and clean way to retrieve it without broader architectural changes.

2. Unsupported dtypes are silently skipped
If a dtype (e.g., "category") is not mappable to a PyArrow type via pa.from_numpy_dtype, we currently drop it from column_types. This fallback behavior avoids breaking the pipeline, but it may lead to silent mismatches when PyArrow falls back to its default inference. We may want to revisit this to either emit warnings or fail explicitly for better visibility.

3. Category dtype handling is not implemented
Right now, category columns are explicitly skipped. We'd need a more nuanced mapping strategy (perhaps via Arrow extension types or metadata) to support this properly.

4. Possible redundancy in _finalize_dtype()
Now that dtypes are mapped earlier during the pyarrow conversion, we may no longer need the final call to self._finalize_dtype(). However, it might still be necessary for preserving native pandas types (e.g., CategoricalDtype) in some edge cases. Open to feedback on whether this should stay.

jbrockmendel commented:
> any additional comments after review?

Only that it looks like you're still using AI in which case a careful review isn't a good use of my time.

dxdc committed:
Merge branch 'main' into patch-2

dxdc commented:
@jbrockmendel I've put significant effort into understanding this bug and implementing the fix. While I used various resources for reference (as most developers do), this is my work and I stand behind it. Happy to discuss any specific concerns about the implementation.

dxdc committed:
Merge branch 'main' into patch-2

dxdc committed:
Merge branch 'main' into patch-2

dxdc commented:
> Known Issues / Remaining Work
Global dtype support is not implemented
The current logic handles only column-specific dtype dictionaries. Global dtypes (e.g., dtype=str) are ignored, which could lead to inconsistent behavior across engines, especially with things like leading-zero string preservation. Supporting this would require column name/index context, which doesn't seem to be readily available here. I couldn't find a safe and clean way to retrieve it without broader architectural changes.

PR for Arrow here, which addresses this on the arrow side: https://github.com/apache/arrow/pull/47663

If approved, we can integrate global dtype via `default_column_type` parameter.

jbrockmendel commented:
> If approved, we can integrate global dtype via default_column_type parameter.

Eventually.  We support older versions of pyarrow for a year (i think)

jbrockmendel commented on pandas/io/parsers/arrow_parser_wrapper.py:
```diff
@@ -139,6 +141,30 @@ def handle_warning(invalid_row) -> str:
+                        if target_dtype:
+                            column_types[col] = target_dtype
+
+                    except TypeError:
```
whats an example where this happens?

jbrockmendel commented on pandas/io/parsers/arrow_parser_wrapper.py:
```diff
@@ -139,6 +141,30 @@ def handle_warning(invalid_row) -> str:
+            else:
+                # TODO: Global dtypes not supported - may cause inconsistent behavior
+                # between engines, especially for leading zero preservation
+                pass
```
if they pass a singleton, can we do something like

`convert_options["column_types"] = defaultdict(user_passed_dtype)`

?

dxdc replied:
It's a good thought, but it doesn't work. One of the first things I tried actually :) I documented a larger analysis on pyarrow here: https://github.com/apache/arrow/issues/47502

You can see the relevant portion of pyarrow code here. Everything is mapped back to C++, and if the column name is not found, it uses the default option.

https://github.com/apache/arrow/blob/eb9d5194a306f8145f8600b176f3bd391ee4397c/cpp/src/arrow/csv/reader.cc#L675-L682

dxdc committed:
remove TypeError check

dxdc committed:
Merge branch 'main' into patch-2

jbrockmendel commented:
Closing as no-action. Can reopen if @jorisvandenbossche wants to weigh in.

jorisvandenbossche commented:
Sorry about that, our new automation to prevent multiple people to open PRs for the same issue kicked in unnecessarily here. Assigned the issue to you and will reopen again.

jorisvandenbossche commented:
The actual logic does not seem to be that complex to me, and if we want to properly support the pyarrow engine, I agree we should try to pass down user-specified `dtype` argument (otherwise I think we should at least disallow it for the pyarrow engine, instead of silently giving wrong results in some cases).



> 2\. Unsupported dtypes are silently skipped
>         If a dtype (e.g., "category") is not mappable to a PyArrow type, we currently drop it from column_types. This fallback behavior avoids breaking the pipeline, but it may lead to silent mismatches when PyArrow falls back to its default inference. We may want to revisit this to either emit warnings or fail explicitly for better visibility.

The `category` case could actually be fixed by trying the convert the `cat_dtype.categories.dtype` to the pyarrow dtype.

Can you think of other cases where this is not possible? 
A potential fix would be to read those columns as strings, so that the string->requested dtype conversion happens fully in the post-processing on the pandas side.



> 3\. Possible redundancy in _finalize_dtype()
>         Now that dtypes are mapped earlier during the pyarrow conversion, we may no longer need the final call to self._finalize_dtype(). However, it might still be necessary for preserving native pandas types (e.g., CategoricalDtype) in some cases that do not have native pyarrow support.

I think it is fine to still have that running always. If we already have the requested dtype, I expect that the logic in `_finalize_dtype` should be cheap (might be worth to check though)

dxdc commented:
Thanks for reopening this @jorisvandenbossche. Just pushed an update based on your suggestions, and integrated handling for the upstream change in pyarrow.

> we should try to pass down user-specified dtype argument

Done for string dtypes, object, and category - the cases where inference actually loses information. I left numeric/bool/datetime on the infer-then-cast path on purpose: casting an inferred column to a numeric target is lossless, and pushing them down breaks parity with the other engines. e.g. "True"/"False" with dtype=float gets inferred as bool then cast by the c/python engines, but pyarrow rejects it at parse time. test_bool_and_nan_to_float / test_bool_and_nan_to_int fail if you push everything down (I tried).

> The category case could actually be fixed by trying the convert the cat_dtype.categories.dtype to the pyarrow dtype. Can you think of other cases where this is not possible? A potential fix would be to read those columns as strings

Implemented both. Added a _pyarrow_parse_type helper: str/"string", object, and bare "category" are read as strings, so the string -> requested dtype conversion happens on the pandas side. This also fixes dtype=object and dtype="category" to match the other engines - six previously-xfailed pyarrow tests now pass, including the #9435 index_col case. For CategoricalDtype with explicit categories I used your categories.dtype idea: string categories read as strings, CategoricalDtype([1, 2]) etc. stays on inference where astype already works. Other unmappable cases (datetime64 with unit, timedelta, nullable EAs) are all numeric-like, so infer-then-cast is already lossless there and I don't think we need a warning anymore.

> I expect that the logic in _finalize_dtype should be cheap (might be worth to check though)

Checked - astype to an identical dtype is ~free, ~1ms on a 1M row dtype=str read vs ~450ms for the read. Left it running always since it still handles the inference-path dtypes, categoricals, and the scalar expansion that skips parse_dates columns.

Also: arrow merged default_column_type (apache/arrow#47663), out in pyarrow 25.0, so scalar dtype=str works now too. Gated behind pa_version_under25p0 - older pyarrow keeps current behavior with a conditional xfail on the scalar tests, dict path works everywhere.

dxdc commented:
edit: @jorisvandenbossche (sorry: wrong tag initially)

jorisvandenbossche commented:
> I left numeric/bool/datetime on the infer-then-cast path on purpose

Ah, yes, especially the bool True/False issue is a good reason to keep that for our post-processing

jorisvandenbossche commented on pandas/io/parsers/arrow_parser_wrapper.py:
```diff
@@ -183,6 +187,33 @@ def handle_warning(invalid_row) -> str:
+            if isinstance(self.dtype, dict):
+                # for a defaultdict only the explicitly-listed columns are
+                # mapped here; the default is applied in read()
+                column_types = {}
+                for col, col_dtype in self.dtype.items():
+                    if not isinstance(col, str):
+                        # pyarrow's column_types requires string column
+                        # names; other labels are cast in _finalize_dtype
+                        continue
```
Do we have a test that ran into this? (as a reason to add this)

With CSV, I would think you can also only have string columns (we don't do dtype inference on the column names in the csv file AFAIK). 
So we might want to raise an error instead of silently ignoring?

jorisvandenbossche replied:
Ah, of course one can pass manual `names` instead of reading it from the file, and then the column names _can_ be non-string values.

dxdc replied:
test_dtype_dict_non_string_key_pyarrow hits it directly - without the guard, ConvertOptions raises TypeError on non-string keys.

test_dtype_per_column also passes 1: str through this path.

I don't think we should raise: as you said, non-string labels are legal via names (and positional int keys work on the c engine), and they aren't silently ignored here - _finalize_dtype still applies them after the read, same as before this PR. The remaining gap is that a text dtype keyed by a non-string label can still lose zeros on this engine, but string keys have the same gap whenever names or header=None is used, since pyarrow parses under autogenerated f0/f1 names and nothing matches column_types at parse time. Mapping user labels to the autogenerated names positionally could close both in a follow-up.

jorisvandenbossche commented on pandas/io/parsers/arrow_parser_wrapper.py:
```diff
@@ -221,6 +252,40 @@ def handle_warning(invalid_row) -> str:
             "encoding": self.encoding,
         }
 
+    def _pyarrow_parse_type(self, dtype) -> pa.DataType | None:
```
This doesn't use `self` right? I am fine with keeping it on the class as a helper method, but then maybe make this explicit with marking it as a staticmethod, or otherwise move it as a helper function outside of the class

dxdc replied:
Moved it out of the class as a module-level function.

jorisvandenbossche commented on pandas/io/parsers/arrow_parser_wrapper.py:
```diff
@@ -221,6 +252,40 @@ def handle_warning(invalid_row) -> str:
+        if dtype == object:
+            # match the other engines: object dtype holds the raw strings
+            return pa.string()
+        target_dtype = to_pyarrow_type(dtype.type)
```
Is there a reason to pass `dtype.type` instead of `dtype` here?

dxdc replied:
Yes - to_pyarrow_type doesn't accept extension dtype instances: pa.from_numpy_dtype raises TypeError for e.g. StringDtype, and to_pyarrow_type only catches ArrowNotImplementedError. The scalar type round-trips for everything needed here (str -> pa.string()). Added a comment. Could teach to_pyarrow_type to handle those directly, but that seemed like separate cleanup.

jorisvandenbossche replied:
Updating `to_pyarrow_type` to do that sounds as a good idea for a follow-up, but indeed lets leave it as is here in this PR

jorisvandenbossche reviewed (commented):
Looking good!

According to the docs, the `dtype` argument is also allowed to be a defaultdict. Is that tested (xfailed) with the pyarrow engine?

I assume you could handle that as well in the `if isinstance(self.dtype, dict):` path: check if it is a defaultdict, and if so, get the `self.dtype.default_factory()` and treat that the same as the non-dict dtype.

dxdc commented:
> I assume you could handle that as well in the if isinstance(self.dtype, dict): path: check if it is a defaultdict, and if so, get the self.dtype.default_factory() and treat that the same as the non-dict dtype.

Good idea. default_factory() now routes through the same helper into default_column_type, and explicit keys still go through column_types, which takes precedence over the default on the pyarrow side. Added test_leading_zeros_preserved_with_dtype_defaultdict with the same conditional xfail for pyarrow<25. The existing defaultdict tests (#41574) all use numeric defaults so they stay unchanged.

Btw, the pyarrow nightly test failed due to a 404 error.

jorisvandenbossche commented on pandas/io/parsers/arrow_parser_wrapper.py:
```diff
@@ -183,6 +225,47 @@ def handle_warning(invalid_row) -> str:
+            # text are applied while parsing; see _pyarrow_parse_type
+            if isinstance(self.dtype, dict):
+                # for a defaultdict only the explicitly-listed columns are
+                # mapped here; the default is applied in read()
```
```suggestion
```

That is also explained in the branch for defaultdict below, so don't think this is needed here

dxdc replied:
done

jorisvandenbossche commented on pandas/tests/io/parser/dtypes/test_dtypes_basic.py:
```diff
@@ -825,3 +828,127 @@ def test_index_col_with_dtype_no_rangeindex(all_parsers):
+    tm.assert_frame_equal(result, expected)
+
+
+def test_dtype_dict_non_string_key_pyarrow(pyarrow_parser_only):
```
Is there a reason to make this test to be pyarrow_parser_only? I understand this is something specific for the pyarrow engine that came up in review, but in the end this should still work the same accross engines?

dxdc replied:
reworked to all_parsers as test_dtype_dict_non_string_key. The reason it was pyarrow-only: the engines genuinely diverge on what an integer key does (c/python apply it positionally, pyarrow can only match labels), so with 1: str there was no single expected frame. Changed the key's dtype to 1: "int64" on an already-int column, which makes the outcome identical everywhere while still exercising the guard.

jorisvandenbossche replied:
Ah, in that case, let's actually keep the test as it was. Because there is a genuine behaviour difference that should ideally be resolved (the fact that in case of pyarrow the integer keys are not applied positionally in the post processing), and that in itself is covered by the `test_dtype_per_column` test as you mentioned, and testing it as in your last update seems to test that this behaviour is correct

dxdc replied:
makes sense - reverted to the pyarrow-only version.

jorisvandenbossche commented on pandas/io/parsers/arrow_parser_wrapper.py:
```diff
@@ -43,6 +47,44 @@
+        if dtype.categories is None:
+            # match the other engines: the categories are the raw
+            # strings from the file
+            return pa.string()
```
```suggestion
            return pa.large_string()
```

Since the default string dtype uses pyarrow's "large_string" instead of "string" under the hood, I think it is slighty more efficient to use that directly here (and in the other cases below where this function returns `pa.string()`)

If someone would have specified a different one (eg manually did `dtype=pd.ArrowDtype(pa.string())`), the casting afterwards still corrects it.

dxdc replied:
all three return points now give pa.large_string(), and the final check accepts both is_string and is_large_string (needed: pa.types.is_string(pa.large_string()) is False).

jorisvandenbossche reviewed (commented):
Looks good! Some final small comments

dxdc commented:
@jorisvandenbossche 

While testing this I ran into a few more pyarrow-engine differences that are out of scope here. Noting them so they don't get lost - happy to work on these as follow-ups once this is in (or split them into issues if that's preferred).

**1. dtype dict keys don't resolve when header=None / names is used**

pyarrow parses under autogenerated column names (f0..fN) in that case, so nothing matches `column_types` at parse time: string dtypes fall back to the lossy post-read cast, and positional integer keys (which the c and python engines support) don't work at all on this engine.

```python
pd.read_csv(data, header=None, names=["col1", "col2"], dtype={"col1": str})
# pyarrow: leading zeros lost; c/python: preserved
pd.read_csv(data, header=None, dtype={0: str})
# pyarrow: not applied while parsing
```

The labels are fully known before parsing in this state (names, or the positions themselves), so the keys can be resolved to fN up front - per column, label first then position, which is the same resolution order the c engine uses. I have this working locally, and it un-xfails `test_str_nan_dropped` (#21131), which is exactly this case.

**2. Column labels diverge with header=None + usecols**

Pre-existing and unrelated to dtypes:

```python
data = "1,2,3\n4,5,6"
pd.read_csv(StringIO(data), header=None, usecols=[0, 2], engine="c").columns
# Index([0, 2], dtype='int64')
pd.read_csv(StringIO(data), header=None, usecols=[0, 2], engine="pyarrow").columns
# Index([0, 1], dtype='int64')  (a RangeIndex on pandas 3.x)
```

The c and python engines keep the file positions as labels, pyarrow renames to range(n). Besides the inconsistency itself, it means a dtype dict keyed by file position can never match the output column on this engine.

**3. names + usecols together**

The label -> file position mapping depends on the usecols selection, so even with (1) those columns would still fall back to post-processing. Solvable, just needs the selection-aware mapping.

**4. to_pyarrow_type and extension dtype instances**

As discussed above - it only accepts numpy dtypes and scalar types today (`pa.from_numpy_dtype` raises TypeError for e.g. StringDtype, which isn't caught), which is why this PR passes `dtype.type`. Teaching it to handle extension dtypes directly would simplify the helper here, and the `is_large_string` branch of the check is there so that change can't silently regress string dtypes to the inference path.

jorisvandenbossche commented:
> While testing this I ran into a few more pyarrow-engine differences that are out of scope here.

Thanks for noting them! 

> 1. dtype dict keys don't resolve when header=None / names is used

That should indeed be possible to map to the autogenerated names, feel free to open a PR with the changes you already have (after this is merged)

> 2. Column labels diverge with header=None + usecols

I wonder if we actually have a test about this (a quick check only showed tests where we also specify `names`, and so don't rely on the auto-generated names). 
This is also not really documented how the auto-generated names and `usecols` should interact (i.e. are names generated before or after applying `usecols`). I suppose the behaviour of the C engine makes sense (and since this is the default one, better to not change this one), but might be worth opening an issue first.


---

While testing the integer key in `dtypes`, I also noticed a strange behaviour with the c engine:

```python
>>> data = "01,02\n02,03"
# reading by default parses as numbers
>>> pd.read_csv(StringIO(data), names=["a", "b"]).dtypes
a    int64
b    int64
dtype: object
# but with dtype dict we can override this to stay as string
>>> pd.read_csv(StringIO(data), names=["a", "b"], dtype={"a": str}).dtypes
a      str
b    int64
dtype: object
# however when using integer key, both columns stay as string
>>> pd.read_csv(StringIO(data), names=[1, 2], dtype={1: str}).dtypes
1    str
2    str
dtype: object
```

Maybe because the `1` is both applied as a label and positionally?

dxdc commented:
> This is also not really documented how the auto-generated names and usecols should interact (i.e. are names generated before or after applying usecols). I suppose the behaviour of the C engine makes sense (and since this is the default one, better to not change this one), but might be worth opening an issue first.

Sounds good, will open the header=None PR once this is merged, and I'll file the usecols/auto-generated names question as an issue - agreed the c engine behavior is the one to keep, especially as the default.

On the integer key behavior: yes, it's exactly that. parsers.pyx resolves the dtype per column, name first and then position:


https://github.com/pandas-dev/pandas/blob/4e54dd017d57ae524f74cee9445750c649633809/pandas/_libs/parsers.pyx#L1744-L1751

So with names=[1, 2], the key 1 hits column 0 by name and column 1 by position. The python engine behaves the same way. I ran into it while prototyping the header=None mapping and replicated name-then-position per column there for consistency, but whether a single key should be able to type two columns is a fair question - the docs only describe dict keys as column names, so the positional fallback (let alone the double application) isn't really specified anywhere.

jorisvandenbossche commented:
Thanks @dxdc for the contribution!
